src/dir.rs

```diff
@@ -0,0 +1,167 @@
+use {Error, Result};
+use errno::Errno;
+use libc;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, fmt, ptr};
+
+#[cfg(target_os = "linux")]
+use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+/// An open directory.
+///! This is a lower-level interface than `std::fs::ReadDir`. Notable differences:
+///!    * can be opened from a file descriptor (as returned by `openat`, perhaps before knowing
+///!      if the path represents a file or directory).
+///!    * implements `AsRawFd`, so it can be passed to `fstat`, `openat`, etc.
+///!    * can be iterated through multiple times without closing and reopening the file
+///!      descriptor. Each iteration rewinds when finished.
+///!    * returns entries for `.` (current directory) and `..` (parent directory).
+///!    * returns entries' names as a `CStr` (no allocation or conversion beyond whatever libc
+///!      does).
+pub struct Dir(ptr::NonNull<libc::DIR>);
+
+impl Dir {
+    #[inline]
+    pub fn from<F: IntoRawFd>(fd: F) -> Result<Self> {
+        Dir::from_fd(fd.into_raw_fd())
+    }
+
+    /// Converts from a file descriptor, closing it on success or failure.
+    pub fn from_fd(fd: RawFd) -> Result<Self> {
+        match ptr::NonNull::new(unsafe { libc::fdopendir(fd) }) {
+            None => {
+                let e = Error::last();
+                unsafe { libc::close(fd) };
+                Err(e)
+            }
+            Some(d) => Ok(Dir(d)),
+        }
+    }
+}
+
+// `Dir` is not `Sync`, but it's not reference-counted either, so it can be safely passed from one
```

Comment:
What makes it non-`Sync` ?

Reply:
Hmm, I was under the impression it was only thread-safe when used with different `DIR` objects. It turns out I was wrong with `readdir_r`, but according to [this glibc documentation](https://www.gnu.org/software/libc/manual/html_node/Reading_002fClosing-Directory.html), that call might be obsoleted, and `readdir` will be specified to work as I'd imagined.

I'm inclined to update the comment and leave it non-`Sync`. Opinions?